src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValues.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.vectorvalues;
+
+import lombok.Getter;
+import lombok.ToString;
+import org.apache.lucene.codecs.KnnFieldVectorsWriter;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.ByteVectorValues;
+import org.apache.lucene.index.FloatVectorValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.opensearch.knn.index.codec.util.KNNCodecUtil;
+
+import java.io.IOException;
+
+/**
+ * An abstract class to iterate over KNNVectors, as KNNVectors are stored as different representation like
+ * {@link BinaryDocValues}, {@link FloatVectorValues}, {@link ByteVectorValues}, {@link KnnFieldVectorsWriter} etc.
+ * @param <T>
+ */
+@ToString
+public abstract class KNNVectorValues<T> {
+
+    @Getter
+    KNNVectorValuesIterator vectorValuesIterator;
+    protected int dimension;
+
+    protected KNNVectorValues(final KNNVectorValuesIterator vectorValuesIterator) {
+        this.vectorValuesIterator = vectorValuesIterator;
+    }
+
+    /**
+     * Return a vector reference. If you are adding this address in a List/Map ensure that you are copying the vector first.
+     * This is to ensure that we keep the heap and latency in check by reducing the copies of vectors.
+     *
+     * @return T an array
+     * @throws IOException if we are not able to get the vector
+     */
+    public abstract T getVector() throws IOException;
+
+    /**
+     * Dimension of vector is returned. Do call getVector function first before calling this function otherwise you will get 0 value.
+     * @return int
+     */
+    public int dimension() {
+        return dimension;
+    }
+
+    /**
+     * Returns the total live docs for KNNVectorValues.
+     * @return long
+     */
+    public long totalLiveDocs() {
+        if (vectorValuesIterator instanceof KNNVectorValuesIterator.DocIdsIteratorValues) {
```

Comment:
KNNVectorValuesIterator interface is defined in KNN, so instead of instance checks delegate to its implementation to reduce branching.

```
KNNVectorValuesIterator {
........//existing methods
      default long totalLiveDocs() {
             return getDocsIdSetITerator().cost(); // If you don't like this you can move it to FieldsValueIterator
      }
}
```
```
DocIdsIteratorValues {
    
long totalLiveDocs() {
      if (instance of KNN80BinaryDocValues) { // Yes, don't depend on knncodecutil unless there is a strong need, // this simplifies the logic for you
             return getDocsIdSetITerator().getTotalLiveDocs
      } 
     return getDocsIdSetITerator().cost()
}
}
```

and then this will become
```
public long totalLiveDocs() {
       return vectorValuesIterator.totalLiveDocs();
}
```
I understand that there are certain validations in place here but think through it, is it a dead validation or not. If there is really a need move it to the implementation of vectorvaluesiterator

Reply:
after doing that, we will be just removing 1 level of if condition. But as it is 1 less nesting, I think its fine. and iterator giving live docs make sense as an interface. BTW lucene uses cost/cardinality for this function.